XLA int8 convolution

yongfeng-nv opened the pull request:
1 Allow int8 for cudnnConvolutionForward and cudnnConvolutionBiasActi…
1.1 Add cast instruction to convert int8 bias to float in cudnn_fused_conv_rewriter.
1.2 Add a new transform cudnn_pad_for_integer_convolution to pad input and output
feature maps to multiple of 4 and pad weights, bias, and side inputs accordingly
2 Add sanity checks for configurations according to cudnn document.
2.1 Input/output feature maps must be multiple of 4.
2.2 Only _IMPLICIT_PRECOMP_GEMM is supported.
2.3 All tensor layouts must be NHWC.
3 Set convolution layout constraints to NHWC for integer convolution.
4 Remove checks for same types in convolution shape inference to allow mixing int8.
acitvation with float bias.
5 Refactor cudnn_pad_for_tensor_cores to share code with cudnn_pad_for_integer_convolution.
5.1 Create a pimpl class CudnnConvPadFeatures to host most common code.
5.2 Define callbacks to resolve desired tensor shapes and use CudnnConvPadFeatures in
cudnn_pad_for_tensor_cores.cc and cudnn_pad_for_integer_convolution.cc.
6 Tests
6.1 Add one unit test to cudnn_conv_rewriter_test.cc for integer forward convolution cudnn API replacement.
6.2 Add unit tests for the newly added CudnnPadForIntegerConvolutions.
6.3 Modify tests in cudnn_fused_conv_rewriter_test.cc.
6.3.1 Add S8 to the type list for testing.
6.3.2 Allow integer value scales for S8 tests.
6.3.3 Skip numerical verification for S8 cases till the int8 HLO evaluator visitor has proper semantics.

yongfeng-nv committed:
1 Allow int8 for cudnnConvolutionForward and cudnnConvolutionBiasActivationForward.
1.1 Add cast instruction to convert int8 bias to float in cudnn_fused_conv_rewriter.
1.2 Add a new transform cudnn_pad_for_integer_convolution to pad input and output
feature maps to multiple of 4 and pad weights, bias, and side inputs accordingly
2 Add sanity checks for configurations according to cudnn document.
2.1 Input/output feature maps must be multiple of 4.
2.2 Only _IMPLICIT_PRECOMP_GEMM is supported.
2.3 All tensor layouts must be NHWC.
3 Set convolution layout constraints to NHWC for integer convolution.
4 Remove checks for same types in convolution shape inference to allow mixing int8.
acitvation with float bias.
5 Refactor cudnn_pad_for_tensor_cores to share code with cudnn_pad_for_integer_convolution.
5.1 Create a pimpl class CudnnConvPadFeatures to host most common code.
5.2 Define callbacks to resolve desired tensor shapes and use CudnnConvPadFeatures in
cudnn_pad_for_tensor_cores.cc and cudnn_pad_for_integer_convolution.cc.
6 Tests
6.1 Add one unit test to cudnn_conv_rewriter_test.cc for integer forward convolution cudnn API replacement.
6.2 Add unit tests for the newly added CudnnPadForIntegerConvolutions.
6.3 Modify tests in cudnn_fused_conv_rewriter_test.cc.
6.3.1 Add S8 to the type list for testing.
6.3.2 Allow integer value scales for S8 tests.
6.3.3 Skip numerical verification for S8 cases till the int8 HLO evaluator visitor has proper semantics.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_pad_features.h:
```diff
@@ -0,0 +1,57 @@
+// Zero-pad input/output feature the desired shapes
+// before the HloCustomCallInstruction calling cuDNN convolution and remove the
+// unecessary output after it
+class CudnnConvPadFeatures {
```
This is a pimpl class hosting common code shared by CudnnConvPadForIntegerConvolutions and CudnnConvPadForTensorCores.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_pad_features.cc:
```diff
@@ -0,0 +1,147 @@
+#include "tensorflow/compiler/xla/window_util.h"
+
+namespace xla {
+namespace gpu {
```
Most code in this file are moved from cudnn_conv_pad_for_tensor_cores.cc.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_pad_for_integer_convolutions.cc:
```diff
@@ -0,0 +1,129 @@
+// We still check the increase for debugging purpose.
+static constexpr double kMaxBytesTouchedIncrease = 1.35;
+
+static StatusOr<bool> ResolvePadedShapes(
```
This function can be replaced with a lambda in theory.  But it is difficult to infer the correct return data type from a lambda in this case (returning true/false, while declaring return type as StatusOr<bool>).

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc:
```diff
@@ -73,7 +73,7 @@ class ScratchBufAllocator : public se::ScratchAllocator {
 };
 
-template <typename T>
+template <typename T, typename BIAS_T>
```
Separate BIAS type to allow mixing int8 activation with float bias.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:
```diff
@@ -40,24 +40,35 @@ class CudnnFusedConvRewriterTest : public HloTestBase {
+    string alpha_conv_scalar, alpha_side_input_scalar;
+    for (absl::string_view type : {"f16", "f32", "f64", "s8"}) {
+      if(type == "s8") {
+        alpha_conv_scalar = "2";
```
Need these changes, because things like "s8[] constant(0.999994934)" are illegal.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:
```diff
@@ -40,24 +40,35 @@ class CudnnFusedConvRewriterTest : public HloTestBase {
                   HasSubstr(kCudnnConvBiasActivationForwardCallTarget));
-      EXPECT_TRUE(RunAndCompare(hlo_with_new_type, ErrorSpec{0.01}))
-          << optimized_hlo_string;
+      if (type != "s8") {
```
Skip numerical verification for s8 until the int8 HLO evaluator visitor has proper semantics.

timshen91 replied:
Ideally we should fix the evaluator before checking in this patch. Do you plan to fix it?

yongfeng-nv replied:
Yes, I am working on it.

yongfeng-nv replied:
int8-to-int8 is disallowed and int8-to-float is implemented in the follow-up commit

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter_test.cc:
```diff
@@ -239,19 +250,19 @@ TEST_F(CudnnFusedConvRewriterTest, TestScaledConvAndScaledSideInputWithBias) {
 
 TEST_F(CudnnFusedConvRewriterTest, TestMatchMaxZeroOnly) {
-  // max(0.1, conv(x, w)) shouldn't match.
+  // max(1, conv(x, w)) shouldn't match.
```
Change 0.1 to 1 won't affect the floating-point test cases.  Otherwise, I will make the change only for the s8 test case.

yongfeng-nv commented on tensorflow/compiler/xla/service/shape_inference.cc:
```diff
@@ -1598,11 +1598,6 @@ ShapeInference::InferDegenerateDimensionBroadcastShape(HloOpcode operation,
         batch_group_count, feature_group_count);
   }
 
-  if (!ShapeUtil::SameElementTypeIgnoringFpPrecision(lhs, rhs)) {
```
Remove this check to allow mixing int8 activation with float bias.

yongfeng-nv replied:
To allow int8-to-float convolution.

yongfeng-nv commented on tensorflow/compiler/xla/shape_util.h:
```diff
@@ -269,7 +269,6 @@ class ShapeUtil {
     if (SameElementType(a, b)) {
       return a.element_type();
     }
-    CHECK(SameElementTypeIgnoringFpPrecision(a, b));
```
Remove this check to allow mixing int8 activation with float bias.

yongfeng-nv commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -2904,6 +2935,11 @@ port::Status CudnnSupport::DoConvolve(
   switch (kind) {
     case dnn::ConvolutionKind::FORWARD: {
       SE_RETURN_IF_ERROR(get_fwd_bugs());
+      if (element_type == stream_executor::dnn::kInt8) {
```
This is ugly. A solution to avoid this special case is to overload DoConvolve() to separate int8 from other types as what DoFusedConvolve (see the other caller to `IntgerConvolutionSanityChecks`) does. However, DoConvolve(int8) still need a if or switch statement to special case `dnn::ConvolutionKind::FORWARD`. Basically DoFusedConvolve only needs to dispatch for data types, while DoConvolve needs to double dispatch for both data types and convolution kinds. I am open for suggestions.

timshen91 replied:
I prefer the dispatch to be in a single function body, using `switch`es and `if`s. Multiple function signatures (through overloading or different function names) is very hard to maintain when the parameters change. I cleaned up most of them, but you can still see the leftovers in `stream.h`

yongfeng-nv replied:
Since the sanity check is removed, this change is not necessary.

yongfeng-nv commented on tensorflow/stream_executor/stream.h:
```diff
@@ -444,6 +455,30 @@ class Stream {
+      DeviceMemory<int8> *backward_input_data,
+      ScratchAllocator *scratch_allocator,
+      const dnn::AlgorithmConfig &algorithm_config,
+      dnn::ProfileResult *output_profile_result) { CHECK(false); }
```
Forgot to change these to errors.

yongfeng-nv committed:
Merge branch 'master' into xla-int8-convolution

timshen91 commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -880,6 +895,36 @@ tf_cc_test(
+)
+
+tf_cc_test(
+    name = "cudnn_conv_pad_for_integer_convolutions_test",
```
Is it worth the cost of adding two file (classes) and one extra pass? How about:
* move all three classes into a single .cc file,
* change file/pass name `pad_for_tensor_cores` to `pad_for_performance`.

yongfeng-nv replied:
I want to point out that these two paddings run under different conditions: pad_for_tensor_cores runs only if IsVoltaOrLater, while it doesn't have this condition.  
I created the new utility class, because the padding/slicing code are so handy that it can be reused by more transforms.  
If you think concise code still more preferable in this case, I am happy to take your suggestion.
pad_for_performance may not be a precise name, as pad_for_integer_convolution is not for performance.  How about pad_for_convolution?

yongfeng-nv replied:
I have combined the two padding in one new transform called pad for convolution in the follow-up commit.  I keep the refactoring but reduce the number of files and classes.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc:
```diff
@@ -73,7 +73,7 @@ class ScratchBufAllocator : public se::ScratchAllocator {
 };
 
-template <typename T>
+template <typename T, typename BIAS_T>
```
`BIAS_T` -> `BiasType` per Google C++: https://google.github.io/styleguide/cppguide.html#Type_Names

yongfeng-nv replied:
I will make the change and bookmarked the guide.  Thank you.

yongfeng-nv replied:
Done

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc:
```diff
@@ -216,6 +216,15 @@ StatusOr<std::unique_ptr<HloInstruction>> TryRewriteToCudnnForwardRelu(
+
+  // Cast integer type bias to float to match 
+  // cudnnConvolutionBiasActivationForward.
+  if (primitive_util::IsIntegralType(bias->shape().element_type())) {
```
Should the requirement "biases should be floats" be pushed to the users?

yongfeng-nv replied:
I got the impression that we prefer taking less assumption from user and trying to make the transforms comprehensive, when I asked about which front-end APIs inferring integer convolution.

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -2731,6 +2731,37 @@ port::Status CudnnSupport::DoPrepareForConvolution(
+    const dnn::BatchDescriptor& output_descriptor,
+    const absl::optional<dnn::AlgorithmDesc>& algo_desc) {
+  // Check algorithm.  Only _IMPLICIT_PRECOMP_GEMM is supported.
+  if (algo_desc.has_value() &&
```
This check may fail with future cuDNN versions (which is also sane), right?

yongfeng-nv replied:
That's correct, if cuDNN changes its behavior.  Putting all checks in one function makes maintenance a little easier.  
Actually, the transforms, such as pad_for_integer_convolution and setting layout to NHWC, will also have to update, when cuDNN changes.
If this is a big concern, we need at least one common predicate for the transforms and sanity check as they are based on the same assumption.  Once cuDNN changes in future, we can follow the predicate to make necessary updates.
How do you think?

yongfeng-nv replied:
I added the sanity checks instead of replying on the status from the cuDNN APIs, because I found unsupported algorithm returns success.  I just received a confirmation from the cuDNN team this morning that it is a bug in the document.  I will revisit the sanity checks and whether it is really necessary.

yongfeng-nv replied:
I have removed the sanity check as the CuDNN APIs work as design.  The document had a bug.

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -2731,6 +2731,37 @@ port::Status CudnnSupport::DoPrepareForConvolution(
+        "cudnnConvolutionBiasActivationForward() for int8 is only supported "
+        "with the input and output feature maps being multiple of 4.");
+  }
+ // Check layout.  Only NHWC is supported.
```
Ditto. Does this check work on future cuDNN versions?

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -2731,6 +2731,37 @@ port::Status CudnnSupport::DoPrepareForConvolution(
   return port::Status::OK();
 }
 
+port::Status CudnnSupport::IntegerConvolutionSanityChecks(
```
In the spirit of making cudan_dnn.cc as thin of a wrapper as possible, I think this sanity check should be in XLA, e.g. RunCudnnConvolutionImpl.

yongfeng-nv replied:
I will try to move the sanity checks.

yongfeng-nv replied:
No need for this change.

timshen91 commented on tensorflow/stream_executor/stream.h:
```diff
@@ -444,6 +455,30 @@ class Stream {
+      DeviceMemory<int8> *backward_input_data,
+      ScratchAllocator *scratch_allocator,
+      const dnn::AlgorithmConfig &algorithm_config,
+      dnn::ProfileResult *output_profile_result) { CHECK(false); }
```
What's the purpose of "implementing" a non-supported convolution?

yongfeng-nv replied:
This implementation is for ThenConvolveBackwardDataWithAlgorithm.  Back propagation with integer type is not supported.

However, after I added S8 to RunCudnnConv, RunCudnnConv now calls RunCudnnConvImpl<int8, float>, which calls/requires ThenConvolveWithAlgorithm<int8> (desired) and ThenConvolveBackwardDataWithAlgorithm<int8> (undesired).  

This is the smallest change I can think of to call ThenConvolveWithAlgorithm<int8>.  Otherwise, we may have to refactor RunCudnnConvImpl and RunCudnnConv to do proper double dispatch on data type and convolution type.

yongfeng-nv replied:
I have refactored these code to use template to dispatch for data types and each data type specialization only calls supported convolutions.  We don't need to implement such non-supported convolutions any more.

timshen91 reviewed (commented):
Thanks for the effort!

The patch looks on the right track, but I had some comments w.r.t. moving code to better places.

yongfeng-nv committed:
Enable int8-to-float convolution node
1. Disallow integer input/kernel to integer output convolution for efficient cuDNN mapping.
2. Allow convolution with integer input/kernel and float output/bias/scaling/side.
3. Define output type a separate argument from element type in DoConvolve/.
4. Refactor RunCudnnConvImpl in cudnn_conv_runner.cc to dispatch cudnn function calls based on input/output types in addition to convolution kind.
5. Add cuDNN pad for integer convolution as an additional pass to cuDNN pad for tensor cores and rename CudnnPadForTensorCores to CudnnPadForConvolutions.
6. Separate accumulator type from element type in HloEvaluatorTypedVisitor's HandleConvolution.
6.1. Add AccumulatorType as a new template argument.
6.2. Use it to make HandleConvolution matching cuDNN's semantics for integer inputs.
6.3. Make corresponding changes in hlo_evaluator and hlo_verifier.
7. Add and update tests for cudnn_conv_rewriter, cudnn_fused_conv_rewriter, and cudnn_pad_for_convolutions.

yongfeng-nv committed:
Merge branch 'master' into xla-int8-convolution

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -845,9 +845,23 @@ cc_library(
-    name = "cudnn_conv_pad_for_tensor_cores",
-    srcs = ["cudnn_conv_pad_for_tensor_cores.cc"],
-    hdrs = ["cudnn_conv_pad_for_tensor_cores.h"],
+    name = "cudnn_conv_pad_features",
```
This target shall be removed.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/BUILD:
```diff
@@ -859,10 +873,10 @@ cc_library(
 tf_cc_test(
-    name = "cudnn_conv_pad_for_tensor_cores_test",
-    srcs = ["cudnn_conv_pad_for_tensor_cores_test.cc"],
+    name = "cudnn_pad_for_convolutions_test",
```
The new pad_for_convolution transform combining pad_for_tensor_cores and pad_for_integer_convolutions from the previous commit.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_algorithm_picker.cc:
```diff
@@ -333,9 +333,9 @@ StatusOr<AutotuneResult> CudnnConvAlgorithmPicker::PickBestAlgorithmNoCache(
-  const auto initialize_buffer = [&stream, &result_shape,
-                                  &rng_state](DeviceMemoryBase buffer) {
-    switch (result_shape.element_type()) {
+  const auto initialize_buffer = [&stream](DeviceMemoryBase buffer,
```
Separate the output type from the element type.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc:
```diff
@@ -521,6 +521,14 @@ StatusOr<bool> RunOnInstruction(HloInstruction* conv) {
 
     // If all else fails, try a forward convolution.
     if (CanImplementAsCudnnForwardConv(conv)) {
+      if (primitive_util::IsIntegralType(
```
This is where we disallow int8-to-int8 convolution for CuDNN mapping.
Make it a fatal error, because int8-to -int8 convolution semantics is not implemented.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc:
```diff
@@ -88,6 +88,17 @@ class CudnnConvRewriterTest : public HloTestBase {
     return CudnnConvRewriter().Run(module).ValueOrDie();
   }
 
+  string GetOptimizedHlo(absl::string_view hlo_string) {
```
We need this function to run numerical verification.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter_test.cc:
```diff
@@ -631,6 +642,46 @@ TEST_F(CudnnConvRewriterTest, BackwardInputConvolveConstantFilter) {
+    })");
+  TF_ASSERT_OK_AND_ASSIGN(auto m, ParseAndReturnVerifiedModule(module_str));
+
+//   ASSERT_FALSE(CudnnConvRewriter().Run(m.get()).ValueOrDie());
```
I want to have a test point to lock down the error for in8-to-int8 convolution.  This is not complete.  I need help to find an example.

yongfeng-nv replied:
What I meant to say was just a test.  I was looking for a test example to lock down a designed error message.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc:
```diff
@@ -149,6 +149,14 @@ absl::optional<ConvWithRelu> FindConvWithRelu(HloInstruction* instr) {
     return absl::nullopt;
   }
 
+  // In order to map to cudnnConvolutionBiasAcitvationForward for integer
```
Today, this check can be replaced with an assert, since cudnn_conv_rewriter won't allow int8-to-int8 convolution go through.  Writing code in the current shape to avoid the assumption/dependency about cudnn_conv_rewriter.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -118,7 +123,61 @@ static Status PadConv(HloCustomCallInstruction* conv,
+// status.
+static StatusOr<bool> ResolveAndPad(
+    HloCustomCallInstruction* conv, StatusOr<bool> (*resolve_pad_shapes)(HloCustomCallInstruction* conv,
+                                         std::vector<Shape>* new_input_shapes,
```
The 2nd and 3rd arguments shall be removed.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -211,29 +271,122 @@ static StatusOr<bool> PadForTensorCores(HloCustomCallInstruction* conv) {
     for (HloCustomCallInstruction* conv : GetRelevantConvs(comp)) {
-      TF_ASSIGN_OR_RETURN(bool result, PadForTensorCores(conv));
-      changed |= result;
+      TF_ASSIGN_OR_RETURN(
```
Have to run the inner loops separately.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h:
```diff
@@ -0,0 +1,47 @@
+namespace xla {
+namespace gpu {
+
+// Zero-pad input/output feature the desired shapes
```
Comments will be improved.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions_test.cc:
```diff
@@ -190,6 +190,79 @@ TEST_F(CudnnConvPadForTensorCoresTest, PadInputFeatures3To4) {
                                ShapeUtil::MakeShape(F16, {2, 2, 4, 32})));
 }
 
+TEST_F(CudnnPadForConvolutionsTest, PadIntForwardConvInputChannels) {
```
The followings are test points moved from pad_for_integer_convolutions

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/gpu_layout_assignment.cc:
```diff
@@ -1,4 +1,4 @@
-/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+ /* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Unnecessary diff.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator.cc:
```diff
@@ -187,7 +187,7 @@ HloEvaluator::HloEvaluator(int64 max_loop_iterations)
   typed_visitors_[U64] =
       absl::make_unique<HloEvaluatorTypedVisitor<uint64>>(this);
-  typed_visitors_[S8] = absl::make_unique<HloEvaluatorTypedVisitor<int8>>(this);
+  typed_visitors_[S8] = absl::make_unique<HloEvaluatorTypedVisitor<int8, int32>>(this);
```
The second template argument is for AccumulatorT.  It is f32 in CuDNN.

yongfeng-nv replied:
I think you are right.  Let me confirm.

yongfeng-nv replied:
This change is not necessary.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1008,7 +1008,8 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
   }
 
-  Status HandleConvolution(HloInstruction* conv) override {
+  template <typename InputT, typename AccumulatorT>
```
Use InputT and AccumulateT instead of InputType and AccumulateType to match other type names in this file.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1040,7 +1040,11 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
                             lhs_shape, rhs_shape, conv->feature_group_count(),
                             conv->batch_group_count(), window, dnums));
-    CHECK(ShapeUtil::Compatible(result_shape, inferred_return_shape))
+    CHECK(ShapeUtil::CompatibleIgnoringElementType(result_shape,
```
Need this change for int8-to-float convolution.

yongfeng-nv commented on tensorflow/stream_executor/cuda/cuda_dnn.h:
```diff
@@ -266,7 +266,8 @@ class CudnnSupport : public dnn::DnnSupport {
 
   port::Status DoConvolve(
-      dnn::ConvolutionKind kind, dnn::DataType element_type, Stream* stream,
+      dnn::ConvolutionKind kind, dnn::DataType element_type,
```
Add the output_type to the function.  This change makes me to touch rocm files.

yongfeng-nv commented on tensorflow/stream_executor/cuda/cuda_dnn.h:
```diff
@@ -336,6 +337,20 @@ class CudnnSupport : public dnn::DnnSupport {
       const dnn::AlgorithmConfig& algorithm_config,
       dnn::ProfileResult* output_profile_result) override;
 
+  bool DoFusedConvolve(
```
The new implementation for int8-to-float convolution.

yongfeng-nv commented on tensorflow/stream_executor/dnn.h:
```diff
@@ -1278,31 +1296,32 @@ class DnnSupport {
-      const BatchDescriptor& input_descriptor, DeviceMemoryBase input_data,
-      const FilterDescriptor& filter_descriptor, DeviceMemoryBase filter_data,
-      const BatchDescriptor& output_descriptor, DeviceMemoryBase output_data,
+      ConvolutionKind kind, DataType element_type, DataType output_type,
```
Adding '''DataType output_type''' to allow int8-to-float convolution and also make me touch rocm files.

yongfeng-nv commented on tensorflow/stream_executor/rocm/rocm_dnn.cc:
```diff
@@ -2808,7 +2808,8 @@ static DeviceMemoryBase MaybeTransformLayout(
 
 port::Status MIOpenSupport::DoConvolve(
-    dnn::ConvolutionKind kind, dnn::DataType element_type, Stream* stream,
+    dnn::ConvolutionKind kind, dnn::DataType element_type,
```
Add '''dnn::DataType output_type''' due to the base class interface change.  It is not used in this function.  In order to verify these changes, I have built Tensorflow for ROCM successfully, though I haven't run any ROCM test.

yongfeng-nv reviewed (commented):
This is a follow-up commit to implement the design that we allow only int8-to-float convolution for CuDNN mapping.  A few comments/issues in the previous commit are avoided/addressed:
1. Int8 convolution semantics is defined and used by tests.
2. No need to deal with non-float bias.
3. No need for specialization for int8-to-int8 convolutions.
4. IntegerConvolutionSanityChecks is removed.
5. Combine pad for tensor cores and pad for integer convolutions in a new transform: pad for convolutions.
Other addressed issues have comments with the code changes.

To do: 
1. when a convert-to-int8 node is behind the int8-to-float convolution node, combine them in one cudnn convolution call.
2. when a convert-to-int8 node is behind the bias/scaling/side input, fuse it in to the cudnn convolution call.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc:
```diff
@@ -73,31 +73,80 @@ class ScratchBufAllocator : public se::ScratchAllocator {
   bool allocated_ = false;
 };
 
-template <typename T>
```
There are many code changes in this file, mostly for one refactoring.

The refactoring starts from RunCudnnConvImpl.  RunCudnnConvImpl used to dispatch for convolution kind and wasn't necessary to dispatch for data types.  We now have to dispatch for data types for integer convolutions, because CuDNN integer convolution have multiple input/output type combinations and doesn't support all kinds of convolutions.

After refactoring, RunCudnnConvImpl calls RunCudnnConvInternalImpl specializations to dispatch for data types.  Each RunCudnnConvInternalImpl specialization dispatches for convolution kind and only calls supported ones.

yongfeng-nv replied:
Sure, I will.

yongfeng-nv committed:
Fix merge issues from the previous commit.

yongfeng-nv commented:
@timshen91 @jlebar  I have updated the PR with new commits, implementing the in8-to-float convolution semantics and addressing comments from Tim.  Please let me know your new comments.

yongfeng-nv committed:
Enable numerical verification in cudnn_fused_conv_rewriter_test.

yongfeng-nv committed:
Map int8 output convolution to a single cudnn convolution call in XLA
1.    Add a new traversal to cudnn_fused_convolution_rewriter to fuse clamping and data type conversion nodes for int8-to-int8 convolution.
2.    Update the error message about XLA int8-to-int8 convolution: in addition to a convert instruction, a clamp instruction is also required for int8 output.
3.    Enable int8-to-int8 cudnn call.
4.    Add tests for two cases:
4.1.    Clamp/convert are next to the convolution instruction.
4.2.    Clamp/convert are next to ReLU.
5.    A minor fix to RunCudnnConvInternalImpl in cudnn_conv_runner.cc.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc:
```diff
@@ -526,8 +526,8 @@ StatusOr<bool> RunOnInstruction(HloInstruction* conv) {
             << "The convolution instruction with integer inputs only allows "
-               "float outputs.  Insert a convert "
-               "instruction after the convolution instruction for interger outputs.";
+               "float outputs.  Insert a clamp instruction with range [-128, 127) followed by a convert "
```
At first, we thought a float-to-int8 conversion would be sufficient.  However, because XLA's float-to-int8 conversion doesn't saturate for overflow, we now look for a custom-call-convolution -> clamp(-128, 127) -> Convert-to-int8 pattern.

jlebar replied:
Do you mean that the front-end (e.g. TensorFlow) will emit a custom-call instruction?

yongfeng-nv replied:
No, custom-calls are from cudnn_convolution_rewriter or the first traversal of cudnn_fused_convolution_rewriter.  I mentioned it here, because it is part of the pattern.  After this transform, the clamp/convert are removed and the custom-call is modified to have int8 output.  Later, cudnn convolution is called with int8 output descriptor.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc:
```diff
@@ -286,5 +284,103 @@ StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+}  // namespace
+
+StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
+  TF_ASSIGN_OR_RETURN(bool fused_for_bias, RunFuseBiasSideActivation(module));
```
I added a second traversal to fused the pattern of custom-call-conv -> clamp -> convert.  It handles both pure convolution and convolution with bias/activation/side cases.

I am open to put the new traversal to a separate transform.

yongfeng-nv commented:
@timshen91 Hi Tim, I saw Justin dismissed the review a week ago.  Will you be able to review it?  It's been open for a long time.  Thank you.

timshen91 commented:
Sorry for the excessive delay on review. I was distracted last month, and mostly out this month.

First off, thanks for the growing amount of work done here. I have some minor comments, but I haven't done reviewing. The PR is on the right track.

Also, in general we prefer small PRs that are potentially dependent. This PR looks likely to be divided into multiple PRs:
* stream_executor change
* evaluator change
* convolution runner change
* pattern matcher change and other changes.

Can you split it if it's relatively easy to do?

timshen91 commented:
Sorry for the excessive delay on review. I was distracted last month, and mostly out this month.

First off, thanks for the growing amount of work done here. I have some minor comments, but I haven't done reviewing. The PR is on the right track.

Also, in general we prefer small PRs that are potentially dependent. This PR looks likely to be divided into multiple PRs:
* stream_executor change
* evaluator change
* convolution runner change
* pattern matcher change and other changes.

Can you split it if it's relatively easy to do?

yongfeng-nv commented:
@timshen91 Thank you for the response.  I will try splitting the commit and merging changes with the HEAD in new PRs.  I may have to defer most tests till the last commit.  Meanwhile, please let me know your comments on the current changes.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -118,7 +123,61 @@ static Status PadConv(HloCustomCallInstruction* conv,
+// shall return true, if padding is necessary or false otherwise in addition to
+// status.
+static StatusOr<bool> ResolveAndPad(
+    HloCustomCallInstruction* conv, StatusOr<bool> (*resolve_pad_shapes)(HloCustomCallInstruction* conv,
```
Can you clang-format all code?

yongfeng-nv replied:
Of course.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -211,29 +271,122 @@ static StatusOr<bool> PadForTensorCores(HloCustomCallInstruction* conv) {
+
+  // Integer convolution only
+  if (!primitive_util::IsIntegralType(input_shape.element_type())) {
+    return false;
```
This seems like an error. Maybe return a `InvalidArgument()` status?

yongfeng-nv replied:
We only need to pad for integer convolutions.  Float convolutions don't need padding.  We just let them go.

timshen91 replied:
Yes, but the function name `ResolvePadedShapesForIntegerConvolution` indicates that it only handles integer convolutions, and it's the caller's responsibility to not pass in non-integer ones.

The current `return false;` behavior seems to better match a function name like `TryResolvePadedShapesForIntegerConvolution`.

Same for `ResolvePadedShapesForTensorCore`. Adding a `Try` is sufficient to clear the confusion.

yongfeng-nv replied:
I agree.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -211,29 +271,122 @@ static StatusOr<bool> PadForTensorCores(HloCustomCallInstruction* conv) {
+  // kForward and kForwardActivation only
+  if (kind != CudnnConvKind::kForward &&
+      kind != CudnnConvKind::kForwardActivation) {
+    return false;
```
Ditto - return `Unimplemented()`?

yongfeng-nv replied:
I don't think we shall do sanity checks here, rather letting the executor tell error.  Although we don't support integer convolution w/ algorithms other than these two today, having sanity checks in this transform may cause maintenance burden, once we can support new algorithms in the future.

timshen91 commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.cc:
```diff
@@ -211,29 +271,122 @@ static StatusOr<bool> PadForTensorCores(HloCustomCallInstruction* conv) {
+    changed |=
+        !ShapeUtil::Equal(conv->operand(i)->shape(), new_input_shapes[i]);
+  }
+  if (!changed) {
```
```
if (!changed) {
  VLOG(3) << ...
}
return changed;
```

yongfeng-nv replied:
Make sense.

timshen91 commented on tensorflow/stream_executor/cuda/cuda_dnn.cc:
```diff
@@ -3737,6 +3738,40 @@ bool CudnnSupport::DoFusedConvolve(
+  int cc_major, cc_minor;
+  stream->parent()->GetDeviceDescription().cuda_compute_capability(&cc_major,
+                                                                   &cc_minor);
+  if (cc_major < 6 || (cc_major == 6 && cc_minor < 1)) {
```
Will cudnn return Unsupported error? If so, maybe leave out the sanity check and just propagate cudnn's runtime error?

yongfeng-nv replied:
This check was there before my changes.  I will check with CuDNN team about whether CuDNN will return unsupported error.  For now, I would like to leave it as is.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_conv_rewriter.cc:
```diff
@@ -521,6 +521,14 @@ StatusOr<bool> RunOnInstruction(HloInstruction* conv) {
+      if (primitive_util::IsIntegralType(
+              conv->operand(0)->shape().element_type()) &&
+              conv->shape().element_type() != F32) {
+        LOG(FATAL)
```
I will make the change.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1203,6 +1207,17 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
   }
 
+  Status HandleConvolution(HloInstruction* conv) override {
+    if (primitive_util::IsIntegralType(conv->operand(0)->shape().element_type())) {
```
OK. I will make the change.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1203,6 +1207,17 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
+      // float.
+      return HandleConvolution<int8, int32>(conv);
+    } else {
+      return HandleConvolution<ReturnT, ElementwiseT>(conv);
```
Make sense.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1203,6 +1207,17 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
+      // CuDNN, i.e. input type int8, elementwise type float, and output type
+      // float.
+      return HandleConvolution<int8, int32>(conv);
+    } else {
```
I am not sure about this comment.  I saw your have another comment about the same code:

if (NativeToPrimitiveType<ReturnT>() == conv->operand(0)->shape().element_type()) {
  return HandleConvolution<ReturnT, ElementwiseT>(conv);
}
return Unimplemented(...);

We will need  return HandleConvolution<ReturnT, ElementwiseT>(conv); for supported types, e.g. float.

yongfeng-nv commented on tensorflow/compiler/xla/service/hlo_evaluator_typed_visitor.h:
```diff
@@ -1008,7 +1008,8 @@ class HloEvaluatorTypedVisitor : public DfsHloVisitorWithDefault {
 
-  Status HandleConvolution(HloInstruction* conv) override {
+  template <typename InputT, typename AccumulatorT>
+  Status HandleConvolution(HloInstruction* conv) {
```
Sure.  I will make the change.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/gpu_layout_assignment.cc:
```diff
@@ -57,6 +57,11 @@ HeuristicLayoutAssignment(const HloInstruction* instr,
                       DataLayout::kBatchYXDepth);
 
+  // Integer convolution must use NHWC.
+  if (primitive_util::IsIntegralType(instr->operand(0)->shape().element_type())) {
```
CuDNN supports integer convolution since compute capability 6.1, prior to Volta.
Integer convolution requires NHWC since then. 
Do you ask how to handle earlier chips that do not support integer convolution?  If so, I intend to leave the sanity check to the executor.

yongfeng-nv commented on tensorflow/compiler/xla/service/gpu/cudnn_pad_for_convolutions.h:
```diff
@@ -0,0 +1,47 @@
+// add slice instruction to remove unnecessary output features.
+class CudnnPadForConvolutions : public HloModulePass {
+ public:
+  explicit CudnnPadForConvolutions(bool is_volta_or_later)
```
Sure.  I will make the change.

yongfeng-nv replied:
@timshen91 CudnnPadForConvolutions taking StreamExecutor* makes it difficult for HLO unit test.  Is it possible for a HLO unit test to pass a StreamExecutor to construct a CudnnPadForConvolutions object?

yongfeng-nv commented:
@timshen91 I have created 4 PRs to replace this one: https://github.com/tensorflow/tensorflow/pull/30761,
https://github.com/tensorflow/tensorflow/pull/30762, https://github.com/tensorflow/tensorflow/pull/30771, and https://github.com/tensorflow/tensorflow/pull/30783.  
Most of your comments have been addressed, except one about passing StreamExecutor* to CudnnPadForConvolutions.  Please let me know if you have further comments.  I will close this one, once we both move on to the new PRs.

mellanox-github commented:
Can one of the admins verify this patch?

yongfeng-nv commented:
Close this PR, as we have moved the necessary changes to https://github.com/tensorflow/tensorflow/pull/30762, https://github.com/tensorflow/tensorflow/pull/30771, and https://github.com/tensorflow/tensorflow/pull/30783.
